### PR TITLE
HTTP client library Doxygen design page

### PR DESCRIPTION
**Pull request by sarenameas**
This PR:
- Adds the design page to the Doxygen
  image

- Adds memory sizes for core_http_client.c and http_parser.c
  image

- Adds documentation for HTTP_DO_NOT_USE_CUSTOM_CONFIG
  image

- Adds better grouping for flags specific to HTTPClient_Send, HTTPRequestInfo_t, and HTTPResponse_t.
  image

Please see the doxygen output:
[output.zip](https://github.com/FreeRTOS/coreHTTP/files/5302973/output.zip)

### CODE REVIEW

File: docs/doxygen/pages.dox [hunk 1]
```diff
@@ -19,34 +19,165 @@ Feature of HTTP/1.1 not supported in this library:
...
 </table>
  */
 
 /**
 @page http_design Design
-HTTP Client Library Design
+HTTP Client Library Architecture and Design
+
+This HTTP client library implements a subset of the HTTP/1.1 protocol. It is
+optimized for resource constrained devices and does not allocate any memory.
```
Comment (yourslab) [inline docs/doxygen/pages.dox:74]:
```suggestion
optimized for resource constrained devices and does not allocate any memory on the heap.
```

File: docs/doxygen/pages.dox [hunk 2]
```diff
@@ -19,34 +19,165 @@ Feature of HTTP/1.1 not supported in this library:
...
+
+@section http_transport_interface_blurb Transport Interface
+
+The HTTP client library relies on a transport interface to dissociate itself from
```
Comment (yourslab) [inline docs/doxygen/pages.dox:78]:
```suggestion
The HTTP client library relies on a transport interface to decouple itself from
```

File: docs/doxygen/pages.dox [hunk 3]
```diff
@@ -19,34 +19,165 @@ Feature of HTTP/1.1 not supported in this library:
...
+Another-Header2: another-value2\r\n
+Another-Header3: another-value3\r\n\r\n
+@endcode
+
+The body is passed into the @ref HTTPClient_Send function when the request is
+ready to be sent.
+
+@section http_range_support HTTP Range Requests and Partial Content Responses
+
+This library strongly encourages the use of Range Requests to download a large.
```
Comment (yourslab) [inline docs/doxygen/pages.dox:132]:
```suggestion
This library strongly encourages the use of Range Requests to download large files.
```

File: docs/doxygen/pages.dox [hunk 4]
```diff
@@ -19,34 +19,165 @@ Feature of HTTP/1.1 not supported in this library:
...
+
+@image html httpclient_send_activity_diagram.png width=50%
+
+@section http_response_headers Reading the HTTP Response Headers
+
+Upon a successful return from @ref HTTPClient_Send, the HTTP Response headers
+can be read from the headers found in @ref HTTPResponse_t.pHeaders. The function
+@ref HTTPClient_ReadHeader reads the headers from an HTTPResponse_t.
+@ref HTTPClient_ReadHeader will re-parse the response in @ref HTTPResponse_t.pBuffer
+looking for the header field of interest. Re-parser involves using http-parser
```
Comment (yourslab) [inline docs/doxygen/pages.dox:167]:
```suggestion
looking for the header field of interest. Re-parsing involves using http-parser
```

Comment (sarenameas) [reply docs/doxygen/pages.dox:167]: thanks

File: docs/doxygen/pages.dox (hunk 3 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:132]: thanks

File: docs/doxygen/pages.dox (hunk 2 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:78]: That is better thanks

File: docs/doxygen/pages.dox (hunk 1 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:74]: I am going to put "dynamically allocate any memory" instead. THanks

File: source/include/core_http_config_defaults.h [hunk 5]
```diff
@@ -35,6 +35,24 @@
...
+ * config file core_http_config.h.
+ *
+ * Without the custom config, the HTTP client library builds with default values
+ * of config macros defined in core_http_config_defaults.h file.
+ *
+ * If a custom config is provided, then HTTP_DO_NOT_USE_CUSTOM_CONFIG should not
+ * be defined.
+ */
+#ifdef DOXYGEN
+    #define HTTP_DO_NOT_USE_CUSTOM_CONFIG
```
Comment (yourslab) [inline source/include/core_http_config_defaults.h:52]: Dan didn't like it when I did this in coreJSON. Would it be possible to copy the @brief text directly into the doxygen file?

File: docs/doxygen/pages.dox [hunk 6]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
+The HTTP client library provides the following API to serialize request headers.
+- @ref HTTPClient_InitializeRequestHeaders
+- @ref HTTPClient_AddHeader
+- @ref HTTPClient_AddRangeHeader
+
+An application is expected to create and populate a @ref HTTPRequestInfo_t and
+set a buffer to use for the headers in @ref HTTPRequestHeaders_t.pBuffer. The
+application may first call @ref HTTPClient_InitializeRequestHeaders to populate
+the @ref HTTPRequestHeaders_t with the method, the path, and the host. The
+HTTP request will be serialized to following when HTTPRequestInfo_t.reqFlags is
```
Comment (yourslab) [inline docs/doxygen/pages.dox:98]:
```suggestion
HTTP request will be serialized to the following when HTTPRequestInfo_t.reqFlags is
```

File: docs/doxygen/pages.dox [hunk 7]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
+@section http_range_support HTTP Range Requests and Partial Content Responses
+
+This library strongly encourages the use of Range Requests to download large
+files. Large is defined here to be a file whose total size cannot fit into the
+space currently available in RAM. By downloading a large file using range
+requests the user application can spend time processing that part of the file
+(for example writing to flash), then request the next part of the file. If the
+user application were to request the entire file at once and process it in
+sections from the network, the system is at a high risk for dropping packets.
+Dropped packets cause retransmissions and with many there can be a negative
```
Comment (yourslab) [inline docs/doxygen/pages.dox:143]:
```suggestion
Dropped packets cause retransmissions and with many, there can be a negative
```

File: source/include/core_http_config_defaults.h (hunk 5 above)
Comment (yourslab) [reply source/include/core_http_config_defaults.h:52]: Actually if this file is made solely for documentation, then it is probably fine. I think the concern was having this in the library code itself.

File: source/include/core_http_client.h [hunk 8]
```diff
@@ -108,19 +106,17 @@
...
+ * @defgroup http_response_flags HTTPResponse_t Flags
  * @brief Flags for #HTTPResponse_t.respFlags.
  * These flags are populated in #HTTPResponse_t.respFlags by the #HTTPClient_Send
  * function.
  *
- * - #HTTP_RESPONSE_CONNECTION_CLOSE_FLAG <br>
- *   @copybrief HTTP_RESPONSE_CONNECTION_CLOSE_FLAG
- * - #HTTP_RESPONSE_CONNECTION_KEEP_ALIVE_FLAG
- *   @copybrief HTTP_RESPONSE_CONNECTION_KEEP_ALIVE_FLAG
+ * A flags value can be extracted from #HTTPResponse_t.respFlags with a
```
Comment (yourslab) [inline source/include/core_http_client.h:114]:
```suggestion
 * A flag's value can be extracted from #HTTPResponse_t.respFlags with a
```

File: docs/doxygen/pages.dox (hunk 7 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:143]: Thanks I need the comma before the with also

File: source/include/core_http_config_defaults.h (hunk 5 above)
Comment (sarenameas) [reply source/include/core_http_config_defaults.h:52]: I have not problem having this in pages.txt only.

File: source/include/core_http_client.h (hunk 8 above)
Comment (sarenameas) [reply source/include/core_http_client.h:114]: Thanks

File: docs/doxygen/pages.dox (hunk 6 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:98]: thanks

**Review by yourslab — APPROVED** [review]

File: docs/doxygen/pages.dox [hunk 9]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
+information.
+
+@section http_request_serialization Building an HTTP Request
+
+The HTTP client library provides the following API to serialize request headers.
+- @ref HTTPClient_InitializeRequestHeaders
+- @ref HTTPClient_AddHeader
+- @ref HTTPClient_AddRangeHeader
+
+An application is expected to create and populate a @ref HTTPRequestInfo_t and
```
Comment (dan4thewin) [inline docs/doxygen/pages.dox:94]: a HTTPRe... -> an HTTPRe...

File: docs/doxygen/pages.dox [hunk 10]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
+<METHOD> <PATH> HTTP/1.1\r\n
+User-Agent: <MY-PLATFORM>\r\n
+Host: <SERVER-URL>\r\n
+Connection: keep-alive\r\n
+Another-Header1: another-value1\r\n
+Another-Header2: another-value2\r\n
+Another-Header3: another-value3\r\n\r\n
+@endcode
+
+The request body is passed into the @ref HTTPClient_Send function when the
```
Comment (dan4thewin) [inline docs/doxygen/pages.dox:131]: This is initiated by the user, yes?

File: docs/doxygen/pages.dox [hunk 11]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
-HTTP Client Library Design
+HTTP Client Library Architecture and Design
+
+This HTTP client library implements a subset of the HTTP/1.1 protocol. It is
+optimized for resource constrained devices and does not dynamically allocate any
+memory.
+
+@section http_transport_interface_blurb Transport Interface
+
+The HTTP client library relies on a transport interface to decouple itself from
```
Comment (dan4thewin) [inline docs/doxygen/pages.dox:79]: Can we avoid anthropomorphism?

File: docs/doxygen/pages.dox [hunk 12]
```diff
@@ -12,41 +12,180 @@ of this library include:
...
+Another-Header2: another-value2\r\n
+Another-Header3: another-value3\r\n\r\n
+@endcode
+
+The request body is passed into the @ref HTTPClient_Send function when the
+request is ready to be sent.
+
+@section http_range_support HTTP Range Requests and Partial Content Responses
+
+This library strongly encourages the use of Range Requests to download large
```
Comment (dan4thewin) [inline docs/doxygen/pages.dox:136]: also anthropomorphic

File: docs/doxygen/pages.dox (hunk 9 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:94]: sure

File: docs/doxygen/pages.dox (hunk 10 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:131]: Yes changing to be more specific thanks

Comment (sarenameas) [reply docs/doxygen/pages.dox:131]:
```
The user application may pass a request body into the @ref HTTPClient_Send
function when the request is ready to be sent.
```

File: docs/doxygen/pages.dox (hunk 11 above)
Comment (sarenameas) [reply docs/doxygen/pages.dox:79]: sure
```
For decoupling from the user platform, the HTTP client library uses a 
transport interface.
```

**Review by dan4thewin — APPROVED** [review]

**Review by yourslab — APPROVED** [review]

**Pull request merged**